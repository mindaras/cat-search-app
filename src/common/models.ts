export interface Option {
  text: string;
  value: string | number;
  selected?: boolean;
}

export type Options = Option[];

export interface Filter {
  options: Option[];
  onChange: (option: Option) => void;
}

export type Filters = Filter[];
