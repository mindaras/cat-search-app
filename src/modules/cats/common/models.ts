export interface Cat {
  id: string;
  url: string;
}

export enum Order {
  Random = "RANDOM",
  Ascending = "ASC",
  Descending = "DESC",
}

export interface Breed {
  id: string;
  name: string;
}
