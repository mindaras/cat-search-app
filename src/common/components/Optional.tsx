interface Props {
  if: boolean;
  children: any;
}

const Optional: React.FC<Props> = (props) => {
  if (props.if) {
    return props.children;
  }

  return null;
};

export { Optional };
