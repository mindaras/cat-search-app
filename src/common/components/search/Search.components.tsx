import styled from "styled-components";

interface ContainerProps {
  disabled: boolean;
}

const Container = styled.div<ContainerProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  height: ${(p) => p.theme.sizes.search.height};
  opacity: ${(p) => (p.disabled ? "0.3" : "1")};

  ${(p) => p.theme.media.small} {
    flex-wrap: wrap;
    height: auto;
    margin: 10px 10px 20px;
}
  }
`;

const Input = styled.input`
  max-width: 100%;
  width: 800px;
  border: 1px solid ${(p) => p.theme.colors.gray.medium};
  box-shadow: 0 0 10px 0px rgb(0 0 0 / 5%);
  border-radius: 4px;
  padding: 15px 20px;
  margin: 0 4px;
  font-size: 14px;
  transition: ease 0.3s;

  &:focus {
    border-color: ${(p) => p.theme.colors.outline.focus};
    outline: 0;
    box-shadow: 0 0 0 0.2rem rgb(0 123 255 / 25%);
  }

  ${(p) => p.theme.media.small} {
    width: 100%;
    margin: 20px 0;
  }
`;

export { Container, Input };
