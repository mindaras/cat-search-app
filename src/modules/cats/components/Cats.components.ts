import styled from "styled-components";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );
`;

export { LoaderContainer };
