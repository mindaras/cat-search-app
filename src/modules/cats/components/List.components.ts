import styled from "styled-components";

const Cat = styled.img`
  width: 25%;
  height: calc(100% / 3);
  object-fit: cover;
  border: 4px solid #fff;
  border-radius: 30px;

  ${(p) => p.theme.media.medium} {
    width: 33.33%;
    height: 30vw;
  }

  ${(p) => p.theme.media.small} {
    width: 100%;
    height: 100vw;
  }
`;

interface ContainerProps {
  single: boolean;
}

const Container = styled.div<ContainerProps>`
  position: relative;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );

  ${(p) => {
    if (p.single) {
      return `
            justify-content: center;
            align-items: center;

            ${Cat} {
                width: 40vw;
                height: 40vw;

                ${p.theme.media.medium} {
                    width: 50vw;
                    height: 50vw;
                }

                ${p.theme.media.small} {
                    width: 80vw;
                    height: 80vw;
                }
            }
        `;
    }
  }};

  ${(p) => p.theme.media.medium} {
    height: auto;
  }
`;

const NoDataText = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 24px;
`;

const NoDataContainer = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: calc(100vh - ${(p) => p.theme.sizes.search.height});

  ${Cat} {
    opacity: 0.1;
  }

  ${(p) => p.theme.media.small} {
    height: auto;
  }
`;

export { Cat, Container, NoDataContainer, NoDataText };
