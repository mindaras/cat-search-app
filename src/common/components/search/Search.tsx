import { debounce } from "@common/helpers/debounce";
import { useCallback, useRef, useState } from "react";
import * as Types from "../../models";
import { Dropdown } from "../dropdown/Dropdown";
import { Container, Input } from "./Search.components";

interface Props {
  value?: string;
  filters?: Types.Filters;
  disabled: boolean;
  filtersDisabled: boolean;
  onChange: (value: string) => void;
}

const Search: React.FC<Props> = ({
  value: initialValue,
  filters,
  disabled,
  filtersDisabled,
  onChange,
}) => {
  const intialValueRef = useRef(initialValue || "");
  const [value, setValue] = useState(intialValueRef.current);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debouncedOnChange = useCallback(
    debounce((value: string) => {
      onChange(value);
    }, 300),
    []
  );

  const changeHandler = (value: string) => {
    setValue(value);
    debouncedOnChange(value);
  };

  return (
    <Container disabled={disabled}>
      <Input
        type="text"
        placeholder="Image id..."
        value={value}
        disabled={disabled}
        onChange={(e) => changeHandler(e.target.value)}
      />
      {filters?.map((filter, i) => (
        <Dropdown
          key={i}
          options={filter.options}
          disabled={disabled || filtersDisabled}
          onChange={filter.onChange}
        />
      ))}
    </Container>
  );
};

export { Search };
