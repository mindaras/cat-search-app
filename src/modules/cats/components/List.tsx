import { Optional } from "@common/components/Optional";
import * as Types from "../common/models";
import { Cat, Container, NoDataContainer, NoDataText } from "./List.components";
import pawImage from "@common/images/paw.jpeg";

interface Props {
  cats: Types.Cat[];
}

const List: React.FC<Props> = ({ cats }) => (
  <Container single={cats.length < 2}>
    <Optional if={!!cats.length}>
      {cats.map((cat) => (
        <Cat key={cat.id} src={cat.url} alt="Cat" />
      ))}
    </Optional>
    <Optional if={!cats.length}>
      <NoDataContainer>
        <Cat src={pawImage} alt="Paw" />
        <NoDataText>No data</NoDataText>
      </NoDataContainer>
    </Optional>
  </Container>
);

export { List };
