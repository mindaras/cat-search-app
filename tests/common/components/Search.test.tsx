import renderer from "react-test-renderer";
import { Container } from "@testUtils/components/Container";
import { Search } from "@common/components/search/Search";
import { Filters } from "@common/models";

describe("Search", () => {
  it("shows an input with filters", () => {
    const filters: Filters = [
      {
        options: [
          { text: "ASC", value: "ASC" },
          { text: "DESC", value: "DESC" },
        ],
        onChange: jest.fn(),
      },
      {
        options: [
          { text: "Abyssinian", value: "abyss" },
          { text: "Aegean", value: "aege" },
        ],
        onChange: jest.fn(),
      },
    ];

    const component = renderer.create(
      <Container>
        <Search
          value="some value"
          filters={filters}
          disabled={false}
          filtersDisabled={false}
          onChange={jest.fn()}
        />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows shows disabled state", () => {
    const filters: Filters = [
      {
        options: [
          { text: "ASC", value: "ASC" },
          { text: "DESC", value: "DESC" },
        ],
        onChange: jest.fn(),
      },
      {
        options: [
          { text: "Abyssinian", value: "abyss" },
          { text: "Aegean", value: "aege" },
        ],
        onChange: jest.fn(),
      },
    ];

    const component = renderer.create(
      <Container>
        <Search
          value="some value"
          filters={filters}
          disabled={true}
          filtersDisabled={false}
          onChange={jest.fn()}
        />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows an input with single filter", () => {
    const filters: Filters = [
      {
        options: [
          { text: "ASC", value: "ASC" },
          { text: "DESC", value: "DESC" },
        ],
        onChange: jest.fn(),
      },
    ];

    const component = renderer.create(
      <Container>
        <Search
          value="some value"
          filters={filters}
          disabled={false}
          filtersDisabled={false}
          onChange={jest.fn()}
        />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows an input with no filters", () => {
    const component = renderer.create(
      <Container>
        <Search
          value="some value"
          disabled={false}
          filtersDisabled={false}
          onChange={jest.fn()}
        />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
