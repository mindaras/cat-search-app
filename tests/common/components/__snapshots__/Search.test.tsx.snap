// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search shows an input with filters 1`] = `
<div
  className="sc-jSUZER kpVgXe"
  disabled={false}
>
  <input
    className="sc-gKPRtg dDQQPa"
    disabled={false}
    onChange={[Function]}
    placeholder="Image id..."
    type="text"
    value="some value"
  />
  <div
    className="sc-gswNZR iLSNqH"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="sc-dkrFOg ircyDF"
    >
       
      <div
        className="sc-bcXHqe bYPlzq"
      >
        <svg
          viewBox="0 0 256 256"
        >
          <g
            transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
          >
            <path
              d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
              strokeLinecap="round"
              transform=" matrix(1 0 0 1 0 0) "
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      className="sc-hLBbgP cIvraF"
      open={false}
    >
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        ASC
      </div>
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        DESC
      </div>
    </div>
  </div>
  <div
    className="sc-gswNZR iLSNqH"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="sc-dkrFOg ircyDF"
    >
       
      <div
        className="sc-bcXHqe bYPlzq"
      >
        <svg
          viewBox="0 0 256 256"
        >
          <g
            transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
          >
            <path
              d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
              strokeLinecap="round"
              transform=" matrix(1 0 0 1 0 0) "
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      className="sc-hLBbgP cIvraF"
      open={false}
    >
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        Abyssinian
      </div>
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        Aegean
      </div>
    </div>
  </div>
</div>
`;

exports[`Search shows an input with no filters 1`] = `
<div
  className="sc-jSUZER kpVgXe"
  disabled={false}
>
  <input
    className="sc-gKPRtg dDQQPa"
    disabled={false}
    onChange={[Function]}
    placeholder="Image id..."
    type="text"
    value="some value"
  />
</div>
`;

exports[`Search shows an input with single filter 1`] = `
<div
  className="sc-jSUZER kpVgXe"
  disabled={false}
>
  <input
    className="sc-gKPRtg dDQQPa"
    disabled={false}
    onChange={[Function]}
    placeholder="Image id..."
    type="text"
    value="some value"
  />
  <div
    className="sc-gswNZR iLSNqH"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="sc-dkrFOg ircyDF"
    >
       
      <div
        className="sc-bcXHqe bYPlzq"
      >
        <svg
          viewBox="0 0 256 256"
        >
          <g
            transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
          >
            <path
              d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
              strokeLinecap="round"
              transform=" matrix(1 0 0 1 0 0) "
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      className="sc-hLBbgP cIvraF"
      open={false}
    >
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        ASC
      </div>
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        DESC
      </div>
    </div>
  </div>
</div>
`;

exports[`Search shows shows disabled state 1`] = `
<div
  className="sc-jSUZER QqSIC"
  disabled={true}
>
  <input
    className="sc-gKPRtg dDQQPa"
    disabled={true}
    onChange={[Function]}
    placeholder="Image id..."
    type="text"
    value="some value"
  />
  <div
    className="sc-gswNZR ckUllH"
    disabled={true}
    onClick={[Function]}
  >
    <div
      className="sc-dkrFOg ircyDF"
    >
      ASC
       
      <div
        className="sc-bcXHqe bYPlzq"
      >
        <svg
          viewBox="0 0 256 256"
        >
          <g
            transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
          >
            <path
              d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
              strokeLinecap="round"
              transform=" matrix(1 0 0 1 0 0) "
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      className="sc-hLBbgP cIvraF"
      open={false}
    >
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        DESC
      </div>
    </div>
  </div>
  <div
    className="sc-gswNZR ckUllH"
    disabled={true}
    onClick={[Function]}
  >
    <div
      className="sc-dkrFOg ircyDF"
    >
      Abyssinian
       
      <div
        className="sc-bcXHqe bYPlzq"
      >
        <svg
          viewBox="0 0 256 256"
        >
          <g
            transform="translate(1.4065934065934016 1.4065934065934016) scale(2.81 2.81)"
          >
            <path
              d="M 90 24.25 c 0 -0.896 -0.342 -1.792 -1.025 -2.475 c -1.366 -1.367 -3.583 -1.367 -4.949 0 L 45 60.8 L 5.975 21.775 c -1.367 -1.367 -3.583 -1.367 -4.95 0 c -1.366 1.367 -1.366 3.583 0 4.95 l 41.5 41.5 c 1.366 1.367 3.583 1.367 4.949 0 l 41.5 -41.5 C 89.658 26.042 90 25.146 90 24.25 z"
              strokeLinecap="round"
              transform=" matrix(1 0 0 1 0 0) "
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      className="sc-hLBbgP cIvraF"
      open={false}
    >
      <div
        className="sc-eDvSVe HjOed"
        onClick={[Function]}
      >
        Aegean
      </div>
    </div>
  </div>
</div>
`;
